### PR TITLE
GPU-based FFT + CMake cleanup + linalg::matrix_product

### PR DESCRIPTION
**Pull request by mhaseeb123**
What's new:

- Adding optimized single-source FFT code for CPUs and GPUs
- Cleanup CMake files - Tested with `nvc++/23.7` and `nvc++/23.1` with both `-stdpar=multicore` and `-stdpar=gpu`
- Removed the `--gcc-toolchain` flag from the CMake files to encourage the use of `localrc` file (already setup for `nvc++/23.7` btw or can be done by exporting `GCCLOCALRC=/path/to/localrc`) instead of the bug-prone flag. 
- Added use of `std::experimental::linalg` library for FFT validation.
- Performance analysis for the FFT codes remains until PM GPUs free up for use. Hoping for the best though 🤞🏼
- FIXME: `clang-format` needed on all source and CMake files.

**Note**: FFT apps give a linker error for `libcublas` with `nvc++/23.1`. Added a commented out line in its CMake file which can be uncommented to add the `libcublas` path if needed. Interestingly `nvc++/23.7` does not complain about  `libcublas`.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -171,9 +171,16 @@ endif()
 
 # need to add appropriate flags for stdexec
 set(CMAKE_CXX_FLAGS
-    "${CMAKE_CXX_FLAGS} -stdpar=${STDPAR} -mp=${OMP} --gcc-toolchain=/opt/cray/pe/gcc/12.2.0/bin/ -pthread"
+    "${CMAKE_CXX_FLAGS} -stdpar=${STDPAR} -mp=${OMP} -pthread"
```
Comment (weilewei) [inline CMakeLists.txt:174]: do we need -pthread flag here?

File: CMakeLists.txt [hunk 2]
```diff
@@ -171,9 +171,16 @@ endif()
...
 )
 
+# add -cudalib=cublas if -stdpar=gpu
+if (STDPAR STREQUAL "gpu")
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -DGPUSTDPAR")
+else()
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -UGPUSTDPAR")
```
Comment (weilewei) [inline CMakeLists.txt:181]: can we have a better name here? Also it's not clear to me what's the difference between D and U GPUSTDPAR

File: apps/1d_stencil/stencil_stdexec.cpp [hunk 3]
```diff
@@ -121,12 +128,14 @@ int benchmark(args_params_t const& args) {
             case sch_t::CPU:
                 solution = step.do_work(exec::static_thread_pool(nthreads).get_scheduler(), size, nt);
                 break;
+#if defined(GPUSTDPAR)
```
Comment (weilewei) [inline apps/1d_stencil/stencil_stdexec.cpp:131]: I wonder if we can get rid of this type of macros as they are intrusive to the codebase and we have to maintain the macros.

File: apps/fft/fft-serial.cpp [hunk 4]
```diff
@@ -30,6 +30,67 @@
 
 #include "fft.hpp"
 
+//
+// serial fft function
+//
+std::vector<data_t> fft_serial(data_t *x, const int N, bool debug = false)
```
Comment (weilewei) [inline apps/fft/fft-serial.cpp:36]: suggest to add `[[nodiscard]]` to enforce the fact that we cannot ignore return value from this function

Comment (weilewei) [reply apps/fft/fft-serial.cpp:36]: if we are not updating x, apply `const` for protection.

File: apps/fft/fft-serial.cpp [hunk 5]
```diff
@@ -81,7 +139,10 @@ int main(int argc, char* argv[])
     Timer timer;
 
     // fft radix-2 algorithm
-    fft_serial(y_n.data(), N, N);
+    auto &&y = fft_serial(x_n.data(), N, args.debug);
```
Comment (weilewei) [inline apps/fft/fft-serial.cpp:142]: I wonder is there any reason using r value reference syntax here `auto&&`?  Does it make sense to have `auto const &`?

File: apps/fft/fft-stdexec.cpp [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+ * commons for the fft codes
+ */
+
+#define FFT_STDEXEC
+#include "fft.hpp"
+
+//
+// fft algorithm
+//
+std::vector<data_t> fft(data_t *x, scheduler auto sch, const int N, const int max_threads, bool debug = false)
```
Comment (weilewei) [inline apps/fft/fft-stdexec.cpp:37]: apply `[[nodiscord]]` and `const` where necessary.

File: apps/fft/fft-stdexec.cpp [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+    uint32_t *id = ind.data();
+
+    int shift = 32 - ilog2(N);
+
+    ex::sender auto twiddle = ex::transfer_just(sch, x_r, x, id)
+        | ex::bulk(N, [=](int k, auto x_r, auto x, auto id){
+            id[k] = reverse_bits32(k) >> shift;
+            x_r[k] = x[id[k]];
+        })
+        | ex::then([](auto &&...){});
```
Comment (weilewei) [inline apps/fft/fft-stdexec.cpp:52]: `ex::then([](auto &&...){});` this syntax is new to me, I am curious to learn what does it do here?

File: apps/fft/fft-stdexec.cpp [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+    std::cout << "FFT progress: ";
+
+    // transfer_just sender
+    ex::sender auto tx = ex::transfer_just(sch, x_r);
+
+    for (int k = 0; k < niters; k++, lN*=2)
+    {
+        std::cout << (100.0 * k)/niters << "%.." << std::flush;
+
+        static Timer dtimer;
```
Comment (weilewei) [inline apps/fft/fft-stdexec.cpp:72]: is `dtimer` a debug timer? I saw it is used only in `debug` condition.

File: apps/fft/fft-stdexec.cpp [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+    return x_rev;
+}
+
+//
+// simulation
+//
+int main(int argc, char* argv[])
+{
+    // parse params
+    fft_params_t args = argparse::parse<fft_params_t>(argc, argv);
```
Comment (weilewei) [inline apps/fft/fft-stdexec.cpp:112]: `const`

File: apps/fft/fft-stdexec.cpp [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+
+    if (print_sig)
+    {
+        std::cout << std::endl << "x[n] = ";
+        x_n.printSignal();
+        std::cout << std::endl;
+    }
+
+    // y[n] = fft(x[n]);
+    std::vector<data_t> y;
```
Comment (weilewei) [inline apps/fft/fft-stdexec.cpp:150]: do we want to resize it? it seems it is default un-initialized

File: apps/fft/fft.hpp [hunk 11]
```diff
@@ -266,93 +274,60 @@ class signal
     std::cout << "]" << std::endl;
   }
 
-  bool isFFT(signal &X, int threads = std::thread::hardware_concurrency())
+  bool isFFT(signal &X, scheduler auto sch, int maxN = 20000)
```
Comment (weilewei) [inline apps/fft/fft.hpp:277]: `[[nodiscard]]`

File: apps/fft/fft.hpp [hunk 12]
```diff
@@ -266,93 +274,60 @@ class signal
...
   {
     int N = y.size();
     bool ret = true;
 
-    data_t *Y = new data_t[N];
-    data_t * M  = new data_t[N*N];
-    auto A = std::mdspan<data_t, view_2d, std::layout_right>(M, N, N);
+    //int nparts = N/maxN;
+    //int psize = std::min(N, nparts);
+    //int matsize = psize * psize;
```
Comment (weilewei) [inline apps/fft/fft.hpp:284]: we have commented and used statements here.

File: apps/fft/fft.hpp [hunk 13]
```diff
@@ -266,93 +274,60 @@ class signal
...
-    for (int k = 0; k < lN/2; k++)
-    {
-        x[k] = e[k] + o[k] * WNk(N, k * stride);
-        x[k+lN/2] = e[k] - o[k] * WNk(N, k * stride);
-    }
-
-    return;
-}
\ No newline at end of file
+using sig_t = signal;
```
Comment (weilewei) [inline apps/fft/fft.hpp:333]: a new line is needed for EOF

**Review by weilewei — COMMENTED** [review]
The code looks good overall to me! Thanks for the contribution! 

One major concern is about extensive usage of macros. As our codebase grow, we probably need a better documentation for how does each macro is used. In a more ideal world, less macros, the better.

File: apps/fft/fft.hpp [hunk 14]
```diff
@@ -266,93 +274,60 @@ class signal
...
+      int i = k / N;
+      int j = k % N;
+      F[k] = WNk(N, i*j);
+    });
+
+    // initialize
+    ex::sync_wait(init);
+
+    // compute Y[n] = dft(x[n]) = WNk * x[n]
+    stdex::linalg::matrix_product(std::execution::par, A, mdy, mdY);
```
Comment (weilewei) [inline apps/fft/fft.hpp:307]: I remembered to use stdpar=gpu, we need `par_unseq` execution policy, but I did not find any performance gain at some point. Maybe we can ask Nvidia folks about it.

File: apps/fft/fft-stdexec.cpp [hunk 15]
```diff
@@ -0,0 +1,221 @@
...
+
+    // initialize stdexec scheduler
+    sch_t scheduler = get_sch_enum(sched);
+
+    // launch with appropriate stdexec scheduler
+    switch (scheduler) {
+        case sch_t::CPU:
+            y = fft(x_n.data(), exec::static_thread_pool(max_threads).get_scheduler(), N, max_threads, args.debug);
+            break;
+#if defined(GPUSTDPAR)
```
Comment (weilewei) [inline apps/fft/fft-stdexec.cpp:163]: rename it as `USE_GPU` or similar

File: apps/1d_stencil/stencil_stdexec.cpp (hunk 3 above)
Comment (mhaseeb123) [reply apps/1d_stencil/stencil_stdexec.cpp:131]: done. renamed to `USE_GPU`

File: apps/fft/fft-serial.cpp (hunk 5 above)
Comment (mhaseeb123) [reply apps/fft/fft-serial.cpp:142]: done.

File: apps/fft/fft-stdexec.cpp (hunk 10 above)
Comment (mhaseeb123) [reply apps/fft/fft-stdexec.cpp:150]: fixed.

File: apps/fft/fft.hpp (hunk 12 above)
Comment (mhaseeb123) [reply apps/fft/fft.hpp:284]: removed these

File: apps/fft/fft.hpp (hunk 13 above)
Comment (mhaseeb123) [reply apps/fft/fft.hpp:333]: done!

File: CMakeLists.txt (hunk 1 above)
Comment (mhaseeb123) [reply CMakeLists.txt:174]: removed. was there from before but no longer needed as things compile just fine without it.

**Pull request merged**